hudi-client/src/main/java/org/apache/hudi/index/bloom/HoodieBloomIndexV2.java

```diff
@@ -0,0 +1,321 @@
+  public boolean canIndexLogFiles() {
+    return false;
+  }
+
+  /**
+   * Bloom filters are stored, into the same data files.
+   */
+  @Override
+  public boolean isImplicitWithStorage() {
+    return true;
+  }
+
+  @Override
+  public JavaRDD<WriteStatus> updateLocation(JavaRDD<WriteStatus> writeStatusRDD, JavaSparkContext jsc,
+                       HoodieTable<T> hoodieTable) {
+    return writeStatusRDD;
+  }
+
+  /**
+   * Given an iterator of hoodie records, returns a pair of candidate HoodieRecord, FileID pairs.
+   */
+  class LazyRangeBloomChecker extends
+          LazyIterableIterator<HoodieRecord<T>, List<Pair<HoodieRecord<T>, String>>> {
+
+    private HoodieTable<T> table;
+    private String currentPartitionPath;
+    private Set<String> fileIDs;
+    private IndexFileFilter indexFileFilter;
+    private ExternalSpillableMap<String, BloomFilter> fileIDToBloomFilter;
+    private HoodieTimer hoodieTimer;
+    private long totalTimeMs;
+    private long totalCount;
+    private long totalMetadataReadTimeMs;
+    private long totalRangeCheckTimeMs;
+    private long totalBloomCheckTimeMs;
+    private long totalMatches;
+
+    public LazyRangeBloomChecker(Iterator<HoodieRecord<T>> in, final HoodieTable<T> table) {
+      super(in);
+      this.table = table;
+    }
+
+    @Override
+    protected List<Pair<HoodieRecord<T>, String>> computeNext() {
+      
+      List<Pair<HoodieRecord<T>, String>> candidates = new ArrayList<>();
+      if (inputItr.hasNext()) {
+
+        HoodieRecord<T> record = inputItr.next();
+        try {
+          initIfNeeded(record.getPartitionPath());
+
+          totalMetadataReadTimeMs += hoodieTimer.deltaTime();
+        } catch (IOException e) {
+          throw new HoodieIOException("Error reading index metadata for " + record.getPartitionPath(), e);
+        }
+
+        // <Partition path, file name>
+        Set<Pair<String, String>> matchingFiles = indexFileFilter
+                .getMatchingFilesAndPartition(record.getPartitionPath(), record.getRecordKey());
+
+        totalRangeCheckTimeMs += hoodieTimer.deltaTime();
+
+        matchingFiles.forEach(partitionFileIdPair -> {
+          BloomFilter filter = fileIDToBloomFilter.get(partitionFileIdPair.getRight());
+          if (filter.mightContain(record.getRecordKey())) {
+            totalMatches++;
+            candidates.add(Pair.of(record, partitionFileIdPair.getRight()));
+          }
+        });
+        totalBloomCheckTimeMs += hoodieTimer.deltaTime();
+
+        if (candidates.isEmpty()) {
+          candidates.add(Pair.of(record, ""));
+        }
+      }
+
+      totalCount++;
+      return candidates;
+    }
+
+    @Override
+    protected void start() {
+      totalTimeMs = 0;
+      totalMatches = 0;
+      totalCount = 0;
+      hoodieTimer = new HoodieTimer().startTimer();
+    }
+
+    @Override
+    protected void end() {
+      totalTimeMs = hoodieTimer.endTimer();
+      String rangeCheckInfo = "LazyRangeBloomChecker: "
```

Comment:
I added these for debugging purposes.. We need not really have these checked in per se?

Reply:
yep, please remove an debugging statements and any other metrics that was added for perf tuning.

Reply:
right, we need this. It'll help us to debug index performance issues if someone met in the future, and print only once per partition.

Reply:
metrics are not easy to use, each partition just only print one log, keep it as it is :)